Maximum of N per topology value.

k82cn opened the pull request:
fixes #40358

k82cn commented:
/release-note-none

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41718)
<!-- Reviewable:end -->

smarterclayton commented on pkg/api/types.go:
```diff
@@ -1737,6 +1737,10 @@ type PodAffinityTerm struct {
+	// The max number of replicas per node/topologyKey to pod anti-affinity;
+	// only available for hard pod anti-affinity (RequiredDuringSchedulingIgnoredDuringExecution)
+	// +optional
+	MaxAntiAffinityTolerant int
```
`replicas` is a precise term (things created by a replica set) and doesn't belong here.  If this applies to pods, `MaxPodsPerTopologyKey` would probably be a better description.  What is the effective default for anti-affinity, 1?  Should be documented.

Don't use generic `int`, use `int32`

k82cn commented:
@k8s-bot verify test this

k82cn commented:
@k8s-bot bazel test this

k82cn commented:
@k8s-bot verify test this

k82cn commented:
cc @davidopp , @mwielgus

davidopp commented:
I won't be able to review this for a week or so.

cc/ @wojtek-t (who's also very busy :) for input on scalability implications, if any
cc/ @bsalamat to review whenever he has time

davidopp commented:
(But thank you for doing this.)

k82cn commented:
my pleasure :)

Regards
---
Klaus, Ma (马达), PMP® | Software Architect
Spectrum Computing CfC, STG, IBM GCG
+86-10-8245 4084 | klaus1982.cn@gmail.com <mailto:klaus1982.cn@gmail.com> |  http://t.cn/RMF2x2B <http://t.cn/RMF2x2B>
> On 27 Feb 2017, at 13:05, David Oppenheimer <notifications@github.com> wrote:
> 
> (But thank you for doing this.)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/kubernetes/kubernetes/pull/41718#issuecomment-282629045>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ARRLLZMeBoExY5RYtAe-mQlHEQ6hrnP_ks5rglmYgaJpZM4MFeYd>.
>

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -929,26 +930,32 @@ func (c *PodAffinityChecker) anyPodMatchesPodAffinityTerm(pod *v1.Pod, allPods [
 	for _, existingPod := range allPods {
 		match := priorityutil.PodMatchesTermsNamespaceAndSelector(existingPod, namespaces, selector)
 		if match {
-			matchingPodExists = true
```
What does the second return value mean after your changes? If it is still the same as before, this doesn't look correct to me. Let's assume there is one existing pod that matches the selector on a node and MaxPodsPerTopologyKey is 2. This method used to return true for the second return value in this case, but now it returns false. Is this intended?

I haven't reviewed all the code yet, but I would like to know the answer to this question as it is important to understand the rest of the code.

k82cn replied:
If MaxPodsPerTopologyKey is 2 and only one existing pod matches, it should return false :). 

`anyPodMatchesPodAffinityTerm` is only used for anti-affinity; so if MaxPodsPerTopologyKey not exceed, it's ok (matches is false).

bsalamat replied:
@k82cn That is not true. `anyPodMatchesPodAffinityTerm` is also used for [checking affinity](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L1146). And to be clear, we are talking about the second return value which "indicates whether a matching pod exists anywhere".

k82cn replied:
@bsalamat , re affinity, yes, it's also considered in this PR (MaxPodsPerTopologyKey is 1 for affinity); my bad remember.

> Let's assume there is one existing pod that matches the selector on a node and MaxPodsPerTopologyKey is 2

`podNum > term.MaxPodsPerTopologyKey` is equal to `matchingPodExists`; according to your case (`MaxPodsPerTopologyKey is 2`), there's not enough matched pods, so it return false.

Another case I did not handle well in this PR is the first return value: if failed to `GetNodeInfo` but succeed in next cycle (false in the first cycle, true in the last cycle), we will return true. But good things is that, the first value is used by affinity :).

bsalamat replied:
@k82cn Thanks! Based on what you just mentioned:
> podNum > term.MaxPodsPerTopologyKey is equal to matchingPodExists; according to your case (MaxPodsPerTopologyKey is 2), there's not enough matched pods, so it return false.

Shouldn't `MaxPodsPerTopologyKey` be actually called `MinPodsPerTopologyKey`? It looks like it refers the minimum number of pods we need in order to call this affinity a match.

k82cn replied:
@bsalamat, `MaxPodsPerTopologyKey` is used because we only cover anti-affinity in this case; but maybe `MaxMatchingPods` and `MinMatchingPods` is better: matching pod num is bigger than `MinMatchingPods` and smaller than `MaxMatchingPods`; if affinity term, schedule it, if anti-affinity term, do not schedule it. In this PR, we only cover `MinPodMatchingPods` for anti-affinity.

bsalamat replied:
@k82cn Please look at this function as an independent module, meaning that it should work  as its name and comment imply. The function is called `anyPodMatchesPodAffinityTerm` and it is in fact used for checking both affinity and anti-affinity. Put yourself in the shoes of someone who will read this code in the future. It will be hard for someone to understand the logic.
I think you should write the logic for checking affinity, and use logical negation of the return value for anti-affinity. This is how the current code is written.

k82cn replied:
@bsalamat , yes, it's what I'm trying to do, e.g. no modification to affinity related code. How about following suggestion?

>maybe MaxMatchingPods and MinMatchingPods is better: matching pod num is bigger than MinMatchingPods and smaller than MaxMatchingPods; if affinity term, schedule it, if anti-affinity term, do not schedule it. In this PR, we only cover MinPodMatchingPods for anti-affinity.

davidopp replied:
I didn't read this whole discussion but I will say that unfortunately we've been using "affinity" to mean "affinity and anti-affinity" in some places, and "affinity but not anti-affinity" in others. The top-level struct is called Affinity even though it's affinity and anti-affinity. (And then, confusingly enough, we have NodeAffinity, but both PodAffinity and PodAntiAffinity). We should have called the general concept of "affinity or anti-affinity" something like "PlacementRules." Anyway, now everything is confusing and I'm not sure there's any way out since we can't change the API in a non-backward-compatible way. But, we can try to make variable names and function names and comments as clear as possible under the very unfortunate circumstances.

bsalamat commented:
@k82cn Thanks Klaus for making the changes. The function looks much better now. I will review the rest of the code soon.

k82cn commented:
@k8s-bot bazel test this

k82cn commented:
@k8s-bot cvm gce e2e test this

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1147,6 +1184,60 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
 			return false
 		}
 	}
+
```
Given the new size of the function, what do you think about placing this new block of code in its own function? Something like SatisfiesMaxPodsPerToplogy.

k82cn replied:
Done

bsalamat commented on pkg/api/types.go:
```diff
@@ -1899,6 +1899,11 @@ type PodAffinityTerm struct {
+	// the matching number is larger than minMatchingPods. The default value is 1.
+	// Only available for hard pod anti-affinity (RequiredDuringSchedulingIgnoredDuringExecution).
+	// +optional
+	MinMatchingPods int32
```
How about uint32?

k82cn replied:
Does protobuf support unsigned int?

bsalamat replied:
Yes, as far as I can tell:
https://developers.google.com/protocol-buffers/docs/proto

k82cn replied:
Great, let me have a try :).

k82cn replied:
Done

bsalamat commented on pkg/api/v1/types.go:
```diff
@@ -2115,6 +2115,11 @@ type PodAffinityTerm struct {
+	// the matching number is larger than minMatchingPods. The default value is 1.
+	// Only available for hard pod anti-affinity (RequiredDuringSchedulingIgnoredDuringExecution).
+	// +optional
+	MinMatchingPods int32 `json:"minMatchingPods,omitempty" protobuf:"varint,4,opt,name=minMatchingPods"`
```
uint32?

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -2886,6 +2888,347 @@ func TestInterPodAffinityWithMultipleNodes(t *testing.T) {
+				"nodeA": true,
+				"nodeB": true,
+			},
+			test:   "NodeA and nodeB have same topologyKey. NodeA has an existing pod that match the inter pod anti-affinity rule. The pod's MaxPodsPerTologyKey is 2, the pod can be scheduled onto nodeA, nodeB.",
```
Here and below

nit:s/match/matches
s/MaxPodsPerTopologyKey/MinMatchingPods

k82cn replied:
Done

bsalamat commented on plugin/pkg/scheduler/schedulercache/reconcile_affinity.go:
```diff
@@ -48,6 +48,9 @@ func ReconcileAffinity(pod *v1.Pod) *v1.Affinity {
 			}
 		}
+
+		// Set default value of MaxPodsPerTopologyKey
```
s/MaxPodsPerTopologyKey/MinMatchingPods

k82cn replied:
Done

bsalamat reviewed (changes requested):
Thanks Klaus! The code looks generally good. Most of my comments are nits.

k82cn commented:
@bsalamat , thanks for your comments; I'm going to address them after e2e flaky test :).

aveshagarwal commented:
@kubernetes/sig-scheduling-pr-reviews

k82cn commented on pkg/api/v1/defaults.go:
```diff
@@ -195,6 +195,28 @@ func SetDefaults_PodSpec(obj *PodSpec) {
 }
+func SetDefaults_Affinity(affinity *Affinity) {
+	if affinity != nil {
+		// If pod anti-affinity is not nil, UPDATE default value of MaxPodsPerTopologyKey to 1.
```
Done

k82cn commented on pkg/api/types.go:
```diff
@@ -1899,6 +1899,11 @@ type PodAffinityTerm struct {
 	// for affinity and for RequiredDuringScheduling pod anti-affinity, empty topologyKey is not allowed.
 	// +optional
 	TopologyKey string
+	// The minimal number of pod that matching the selector; the term is considered to be matched if and only if
```
Done

k82cn commented on pkg/api/v1/defaults.go:
```diff
@@ -195,6 +195,28 @@ func SetDefaults_PodSpec(obj *PodSpec) {
+				}
+			}
+		}
+		// If pod affinity is not nil, SET value of MaxPodsPerTopologyKey to 1.
```
Done for the comments.

I left the comments for "perferred", as it's the places I'm going to set default for preferred; so replaced logic with a line of comment.

k82cn commented on pkg/api/v1/types.go:
```diff
@@ -2115,6 +2115,11 @@ type PodAffinityTerm struct {
 	// for affinity and for RequiredDuringScheduling pod anti-affinity, empty topologyKey is not allowed.
 	// +optional
 	TopologyKey string `json:"topologyKey,omitempty" protobuf:"bytes,3,opt,name=topologyKey"`
+	// The minimal number of pod that matching the selector; the term is considered to be matched if and only if
```
Done

k82cn commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -986,30 +987,35 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta interface
 // First return value indicates whether a matching pod exists on a node that matches the topology key,
```
Done

k82cn commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -986,30 +987,35 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta interface
 // First return value indicates whether a matching pod exists on a node that matches the topology key,
 // while the second return value indicates whether a matching pod exists anywhere.
```
Done

k82cn commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -986,30 +987,35 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta interface
+		return false, false, []error{err}
 	}
+
+	matchingPodAnywhere := int32(1)
```
Yes, I changed from 0 to 1 :). 0 makes it means how many OTHER pods that matching the term; so I think 1 is better: how many matching pods in total.

bsalamat replied:
Alright, that is fine.

k82cn commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -986,30 +987,35 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta interface
 		}
 	}
-	return false, matchingPodExists, nil
+	return matchingPodInTopology > term.MinMatchingPods, matchingPodAnywhere > term.MinMatchingPods, errs
```
done

k82cn commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1094,20 +1121,27 @@ func getMatchingAntiAffinityTerms(pod *v1.Pod, nodeInfoMap map[string]*scheduler
 			for _, term := range getPodAntiAffinityTerms(affinity.PodAntiAffinity) {
+				// The MinMatchingPods Terms for anti-affinity is handled by others.
+				if term.MinMatchingPods > 1 {
+					hasMatchingPodsRange = true
```
done

k82cn commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1147,6 +1184,60 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
+		}
+
+		for i := range allPods {
+			existingPod := allPods[i]
```
Done

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -983,33 +984,41 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta interface
 			}
 		}
+		if matchingPodInTopology > term.MinMatchingPods && matchingPodAnywhere > term.MinMatchingPods {
+			break
```
You can `return true, errs` here and `return false, errs` at line 1021.

k82cn replied:
There're two conditions: `matchingPodInTopology > term.MinMatchingPods` && `matchingPodAnywhere > term.MinMatchingPods`; we're not sure which one or both is false; so we can not `return false, false, errs` at line 1021 :).

bsalamat replied:
Oh, sorry, you are right. I missed the comma and thought these were the same condition. Never mind then.

bsalamat reviewed (commented):
Thanks, Klaus. Looks good. Only a couple of minor comments.

k82cn commented:
@k8s-bot unit test this

bsalamat commented:
Awesome! Thanks, Klaus!

bsalamat commented:
/lgtm

k82cn commented:
@wojtek-t , PLTA :).

kevin-wangzefeng commented:
/cc @alfred-huangjian @kubernetes/huawei

k82cn commented:
hi team, please feel free to comment :). The `needs-rebase` is for generated codes; I'll rebase when addressing comments.

k82cn commented:
ping @wojtek-t  :).

k82cn commented:
@k8s-bot non-cri e2e test this

k82cn commented:
rebased; ping @wojtek-t :).

smarterclayton commented on api/swagger-spec/batch_v1.json:
```diff
@@ -3073,6 +3073,10 @@
+     },
+     "minMatchingPods": {
+      "type": "integer",
+      "description": "MinMatchingPods specifies the minimal number of pod that matching the selector; the term is considered to be matched if and only if the matching number is larger than minMatchingPods. The default value is 1. Only available for hard pod anti-affinity (RequiredDuringSchedulingIgnoredDuringExecution)."
```
Godoc on public API fields should use the external field name - minMatchingPods - because it is used in docs, human readable descriptions in the API, etc.

smarterclayton commented:
One more change please before I give API sign off on this.

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -2131,6 +2131,11 @@ type PodAffinityTerm struct {
 	// for affinity and for RequiredDuringScheduling pod anti-affinity, empty topologyKey is not allowed.
 	// +optional
 	TopologyKey string `json:"topologyKey,omitempty" protobuf:"bytes,3,opt,name=topologyKey"`
+	// MinMatchingPods specifies the minimal number of pod that matching the selector; the term is considered to
```
I found the description somewhat confusing. Also, I'm not sure MinMatchingPods is a good name.

The meaning is a little different when it is used with affinity vs. anti-affinity, so I would suggest to write separate comments for each use case.

I didn't look at the implementation yet, but here is what it should be implementing:

```
When used with anti-affinity, a node N is eligible only if the number of pods
matching the LabelSelector that are running on nodes with the same value
for key TopologyKey as N has is less than MinMatchingPods. For example,
if MinMatchingPods is 1 (the default), then after this pod is scheduled to node M,
there will not be more than one pod from the set selected by the LabelSelector
running on nodes with M's value for key TopologyKey.

When used with affinity, a node N is eligible only if the number of pods
matching the LabelSelector that are running on nodes with the same value
for key TopologyKey as N has is greater or equal to MinMatchingPods.
For example, if MinMatchingPods is 1 (the default), then after this pod is
scheduled to node M, there will be at least 2 pods from the set selected by
the LabelSelector running on nodes with M's value for key TopologyKey.
```

(BTW, I actually don't think we need to bother to make this apply to affinity -- the main use case is with anti-affinity. But if it's easier to implement both affinity and anti-affinity, that's OK.)

I would suggest calling it "NumMatchingPods" or something like that. there isn't really a good variable name that captures the full meaning, and anyway this feature is complicated enough that the user is going to need to read the description or documentation anyway.

Please update the comment, change the variable name, and verify that it is implementing what I described above.

k82cn replied:
thanks for your comments :). This PR only implements anti-affintiy; but affinity is easier. I'll update PR accordingly.

davidopp replied:
It's fine if you only handle anti-affinity case (that case is more important than affinity case). But make the comment clear (in the API and the code).

Or if you want to do both affinity and anti-affinity, that's fine too, but same comment applies. :)

k82cn commented:
Updated code by :
1. rebase
2. changed name to `NumOfMatchingPods` and add comments accordingly
3. Affinity is not implemented yet in this PR.

/cc @davidopp  @wojtek-t :).

k82cn commented:
@wojtek-t , do you have review bandwidth to include this in 1.7? If not, I'm ok to move it into next release.

smarterclayton commented:
Let's move this into the next release and give it more attention.

On Fri, Jun 2, 2017 at 6:11 PM, Kubernetes Submit Queue <
notifications@github.com> wrote:

> This PR hasn't been active in 30 days. It will be closed in 59 days (Aug
> 1, 2017).
>
> cc @bsalamat <https://github.com/bsalamat> @davidopp
> <https://github.com/davidopp> @k82cn <https://github.com/k82cn>
> @smarterclayton <https://github.com/smarterclayton> @wojtek-t
> <https://github.com/wojtek-t>
>
> You can add 'keep-open' label to prevent this from happening, or add a
> comment to keep it open another 90 days
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41718#issuecomment-305920843>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p_13oEl0i1xRDiLrdo5D3loQgq1Sks5sAIiEgaJpZM4MFeYd>
> .
>

wojtek-t commented:
> Let's move this into the next release and give it more attention.

Yes, sorry for that.

zhouhaibing089 commented on pkg/api/types.go:
```diff
@@ -1960,6 +1960,24 @@ type PodAffinityTerm struct {
+	//     the LabelSelector running on nodes with M's value for key TopologyKey.
+	// The default value is 1.
+	// +optional
+	NumOfMatchingPods uint32
```
does it make more sense to use `*uint32`? reading the current code, if user explicitly set this field as `0`, it will be override later to be `1`(int the Set Defaults function), while in my understanding, it is more appropriate to raise a validation error rather than modifying it internally.

moving to pointer type, we can set the default value as `1` if it is nil after decoding which seems better to me.

zhouhaibing089 replied:
ping @k82cn

k82cn replied:
Any case that user need to set 0?

zhouhaibing089 replied:
I am not saying 0 is a valid case, just a general comment, the zero value for an uint32 is 0, while the zero value for an *uint32 is nil, you do have some defaulting logic: if it is less than 1, set it to 0, and thus I think it should better to be *uint32.

As an example, the [GracePeriodSeconds](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L3293) has the default value 30s, and pointer type.

k82cn replied:
if 0 is a valid case, we use pointer to distinguish default value and customer setting. In this issue, I think uint32 is enough.

zhouhaibing089 replied:
`0` might be valid case for affinity which means a pod can be scheduled on a node without consideration on the number of pods running on that node.

k82cn commented:
@smarterclayton , @wojtek-t would you help to review this PR? We'd like to get this merged in 1.8 :).

k82cn commented:
/retest

Commit:
Maximum of N per topology value.

k82cn commented:
/retest

zhouhaibing089 commented on pkg/api/v1/defaults.go:
```diff
@@ -176,6 +176,31 @@ func SetDefaults_PodSpec(obj *v1.PodSpec) {
+	if obj.PodAffinity != nil {
+		for i := range obj.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			term := &obj.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution[i]
+			term.NumOfMatchingPods = 1
```
I am not sure this is correct to set this as `1` for affinity unconditionally.

k82cn commented:
/retest

k82cn commented:
@wojtek-t @davidopp , do you have time to review this one?

davidopp commented:
@bsalamat and I were discussing this PR briefly today in the context of preemption. We think supporting this feature for pod affinity may be very difficult. (pod anti-affinity is fine.) Can you remove all of the text about that case (e.g. all of the "When used with affinity (not implemented yet), a node N is eligible only if the number of pods matching the LabelSelector that are running on nodes with the same value for key TopologyKey as N has is greater or equal to NumOfMatchingPods. For example, if NumOfMatchingPods is 1 (the default), then after this pod is scheduled to node M, there will be at least one pods from the set selected by the LabelSelector running on nodes with M's value for key TopologyKey."), and have validation ensure that the user does not set the value > 1 with pod affinity? We could consider implementing it for pod affinity in the future once we understand the implications for preemption better.

k82cn commented:
sure, I'll remove them. currently, this PR only include anti-affinity here.

On Tue, Aug 8, 2017 at 2:43 PM David Oppenheimer <notifications@github.com>
wrote:

> @bsalamat <https://github.com/bsalamat> and I were discussing this PR
> briefly today in the context of preemption. We think supporting this
> feature for pod affinity may be very difficult. (pod anti-affinity is
> fine.) Can you remove all of the text about that case (e.g. all of the
> "When used with affinity (not implemented yet), a node N is eligible only
> if the number of pods matching the LabelSelector that are running on nodes
> with the same value for key TopologyKey as N has is greater or equal to
> NumOfMatchingPods. For example, if NumOfMatchingPods is 1 (the default),
> then after this pod is scheduled to node M, there will be at least one pods
> from the set selected by the LabelSelector running on nodes with M's value
> for key TopologyKey."), and have validation ensure that the user does not
> set the value > 1 with pod affinity? We could consider implementing it for
> pod affinity in the future once we understand the implications for
> preemption better.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41718#issuecomment-320865267>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARRLLczpQlqrTLSI2rM3_siyFTEV-mZNks5sWAORgaJpZM4MFeYd>
> .
>
-- 

Regards,
----
Da (Klaus), Ma (马达), PMP® | Software Architect
IBM Platform Development & Support, STG, IBM GCG
+86-10-8245 4084 | madaxa@cn.ibm.com | http://k82.me

k82cn commented:
@wojtek-t , would you help to review the non generated part?

I'm finding a host to generated them because of network, docker version ...

davidopp commented:
@wojtek-t

k82cn commented:
/retest

zhouhaibing089 commented:
@k82cn it fails the `roundtrip_test.go`, you will have to add some fuzz code at here to make sure `NumOfMatchingPods`  is larger than 0: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/fuzzer/fuzzer.go#L88.

k82cn commented:
/retest

k82cn commented:
/keep-open

bsalamat commented:
@k82cn What is the status of this PR? Is it ready for another round of review?

k82cn commented:
@bsalamat , I had addressed David@'s comments; let me rebase it before another round review. I'd like to get this merge in 1.9.

bsalamat commented:
> I had addressed David@'s comments; let me rebase it before another round review. I'd like to get this merge in 1.9.

SGTM

k82cn commented:
/keep-open

I'll take some time to rebase this :).

jinzhejz commented:
/retest

anfernee commented on pkg/apis/core/fuzzer/fuzzer.go:
```diff
@@ -90,6 +90,27 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
+					},
+				}
+			}
+			if s.Affinity.PodAntiAffinity != nil {
```
this is always true, right?

anfernee commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -186,6 +186,31 @@ func SetDefaults_PodSpec(obj *v1.PodSpec) {
+	if obj.PodAffinity != nil {
+		for i := range obj.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			term := &obj.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution[i]
+			term.NumOfMatchingPods = 1
```
why it overwrites the value not setting only when it's 0, like anti-affinity?

k82cn commented:
/kind feature
/sig scheduling
/priority important-soon

Commit:
generated codes.

bsalamat commented:
@k82cn What is the status of this PR? Are you still working on it?

wojtek-t commented:
@bsalamat @k82cn @davidopp  - I was thinking a bit about this and before we proceed with this PR we should discuss the usecases.
So the case that I have in mind is that I would like to schedule 9 pods across 3 zones. If I want to have equal spreading, I need to specify max replicas to 3. What if I want to increase the number of replicas? Do I always have to adjust the pod affinity on any scale ups/scale downs?

I mean I'm not really convinced about real benefit to the user. I would like to think more about end user experience before introducing this feature. Pod affinity/antiaffinity is already causing us different problems and I would like to avoid making it even more problematic.

k82cn commented:
@bsalamat , yes, waiting for review if anyone have cycle for this.

> So the case that I have in mind is that I would like to schedule 9 pods across 3 zones. If I want to have equal spreading, I need to specify max replicas to 3. 

@wojtek-t , the case is that: in a zone, the user would like to place more Pods for robustness instead of only one pod in each zone, as there're several nodes.

wojtek-t commented:
> @wojtek-t , the case is that: in a zone, the user would like to place more Pods for robustness instead of only one pod in each zone, as there're several nodes.

I understand that. My question is. I wanted equal spreading across zones to I set maxReplicas=3 to place 9 pods. But now instead of 9 pods, I want to have 18. What should I do? Or now I need only 3 replicas. What's then?

What I'm trying to say is that I don't really believe in the usecase that number of replicas will be constant for the whole lifetime of the application. We should ensure that we have some story for for scaling up/down the deployment (e.g. make it somehow work with HPA).

k82cn commented:
> I wanted equal spreading across zones to I set maxReplicas=3 to place 9 pods. But now instead of 9 pods, I want to have 18. What should I do? Or now I need only 3 replicas. What's then?

By this PR, we have to update `maxReplicas` accordingly. For spreading across zones, I'm thinking whether `SelectorSpreadPriority` (by zone) will help; and it seems `SelectorSpreadPriority` by zone will address the requirement. Not sure whether any requirement to set hard max restriction :).

bsalamat commented:
I think @wojtek-t's main point is that we shouldn't add more complexity to our scheduling logic without having practical use-cases. I agree with his point that if this feature is not going to support realistic use-cases, we shouldn't add it.
@k82cn, are there any use-cases or user requests that cannot be addressed with the existing scheduler features and require having this new feature?

wojtek-t commented:
> I think @wojtek-t's main point is that we shouldn't add more complexity to our scheduling logic without having practical use-cases. I agree with his point that if this feature is not going to support realistic use-cases, we shouldn't add it.
@k82cn, are there any use-cases or user requests that cannot be addressed with the existing scheduler features and require having this new feature?

I guess that's one thing. But I would go step further from it. Which is:
"even if we have usecases for this feature, don't we have (equally?) important usecases that this features won't solve and we would need something else that would also solve this usecase".

One potential example that I can imagine is: "I want to have my app equally spreaded across zones". I think this feature doesn't solve this problem, and solution to this problem may potentially solve the problem solved by this PR.

k82cn commented:
> are there any use-cases or user requests that cannot be addressed with the existing scheduler features and require having this new feature?

@bsalamat , please refer to #40358 's description; it's a reasonable user case to me :).

> "I want to have my app equally spreaded across zones"

@wojtek-t , agree, that's better description to the requirement; and this PR did not address the requirement well. Regarding "spreaded across zones", should it be hard (predicate) or soft (priority)?

wojtek-t commented:
> @wojtek-t , agree, that's better description to the requirement; and this PR did not address the requirement well. Regarding "spreaded across zones", should it be hard (predicate) or soft (priority)?

Good question. I think there are good usecases for both. But I would need to think a bit more about it.

bsalamat commented:
> @bsalamat , please refer to #40358 's description; it's a reasonable user case to me :).

@k82cn, I saw that issue before asking the question, but I guess anti-affinity predicate/priority function address the cases mentioned in that issue. BTW, in most Kubernetes deployments, the concept of "rack" is not present and users cannot control that.

k82cn commented:
> I guess anti-affinity predicate/priority function address the cases mentioned in that issue

nop, the anti-affinity predicate makes sure only one pod in each "zone" which introduced concern on robustness.

bsalamat commented:
> nop, the anti-affinity predicate makes sure only one pod in each "zone" which introduced concern on robustness.

Yes, I understand, but at the same time I also think the issue that @wojtek-t mentioned exists if one tries to place more than one Pod per topology and that issue may cause this feature that we are adding in this PR not get used.

k82cn commented:
yes. I think wojtek-t@ 's description of this requirement is better :).

k82cn commented:
@bsalamat , @wojtek-t , I'm going to close this PR and continue the discussion at issue :)
